tool/tsh/resolve_default_addr.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+type raceResult struct {
+	addr string
+	err  error
+}
+
+// nonOKResponseError indicates that the racer made contact with a server &
+// issued a request but received a non-OK response. This is still
+// considered a failure by the port resolution algorithm.
+type nonOKResponseError struct {
+	Status int
+}
+
+func (err nonOKResponseError) Error() string {
+	return fmt.Sprintf("Non-OK response status: %03d", err.Status)
+}
+
+// raceRequest drives an HTTP request to completion and posts the results back
+// to the supplied channel.
+func raceRequest(ctx context.Context, cli *http.Client, addr string, results chan<- raceResult) {
+	target := fmt.Sprintf("https://%s/", addr)
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
+
+	if err == nil {
+		var rsp *http.Response
+		rsp, err = cli.Do(request)
+		if err == nil {
+			rsp.Body.Close()
+
+			// If the request returned a non-OK response then we're still going
+			// to treat this as a failure and return an error to the race
+			// aggregator.
+			if rsp.StatusCode != http.StatusOK {
+				err = nonOKResponseError{Status: rsp.StatusCode}
```

Comment:
If you just send a `trace.BadParameter("received %v response status")` then you wouldn't need extra nonOKResponseError type. Also, maybe log the response body in this case (debug), for troubleshooting?

Reply:
Done